File: function-calling/api-as/assembly/tool-helper.ts
import {
  OpenAIChatModel,
  Tool,
  ToolCall,
  SystemMessage,
  UserMessage,
  ToolMessage,
  ResponseFormat,
  CompletionMessage,
  ToolChoice,
} from "@hypermode/modus-sdk-as/models/openai/chat"

/**
 * Final response and log of each prompt iteration with tool use
 */
@json
export class ResponseWithLogs {
  response: string = ""
  logs: string[] = []
}

export function llmWithTools(
  model: OpenAIChatModel,
  tools: Tool[],
  system_prompt: string,
  question: string,
  toolCallBack: (toolCall: ToolCall) => string,
  limit: u8 = 3,
): ResponseWithLogs {
  var logs: string[] = []
  var final_response = ""
  var tool_messages: ToolMessage<string>[] = []
  var message: CompletionMessage | null = null
  var loops: u8 = 0
  // we loop until we get a response or we reach the maximum number of loops (3)
  do {
    message = getLLMResponse(model, tools, system_prompt, question, message, tool_messages)
    /* do we have a tool call to execute */
    if (message.toolCalls.length > 0) {
      for (let i = 0; i < message.toolCalls.length; i++) {
        logs.push(
          `Calling function : ${message.toolCalls[i].function.name} with ${message.toolCalls[i].function.arguments}`,
        )
      }
      tool_messages = aggregateToolsResponse(message.toolCalls, toolCallBack)
      for (let i = 0; i < tool_messages.length; i++) {
        logs.push(`Tool response    : ${tool_messages[i].content}`)
      }
    } else {
      final_response = message.content
      break
    }
  } while (loops++ < limit - 1)

  return { response: final_response, logs: logs }
}

function getLLMResponse(
  model: OpenAIChatModel,
  tools: Tool[],
  system_prompt: string,
  question: string,
  last_message: CompletionMessage | null = null,
  tools_messages: ToolMessage<string>[] = [],
): CompletionMessage {
  const input = model.createInput([new SystemMessage(system_prompt), new UserMessage(question)])
  /*
   * adding tools messages (response from tools) to the input
   * first we need to add the last completion message so the LLM can match the tool messages with the tool call
   */
  if (last_message != null) {
    input.messages.push(last_message.toAssistantMessage())
  }
  for (var i = 0; i < tools_messages.length; i++) {
    input.messages.push(tools_messages[i])
  }

  input.responseFormat = ResponseFormat.Text
  input.tools = tools

  input.toolChoice = ToolChoice.Auto //  Auto, Required, None, or Function("name")

  const message = model.invoke(input).choices[0].message
  return message
}

/**
 * Execute the tool calls and return an array of ToolMessage
 * containing the response of the tools
 */
function aggregateToolsResponse(
  toolCalls: ToolCall[],
  toolCallBack: (toolCall: ToolCall) => string,
): ToolMessage<string>[] {
  var messages: ToolMessage<string>[] = []
  for (let i = 0; i < toolCalls.length; i++) {
    const content = toolCallBack(toolCalls[i])
    const toolCallResponse = new ToolMessage(content, toolCalls[i].id)
    messages.push(toolCallResponse)
  }
  return messages
}

File: chat-agent/api-go/tool-helper.go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/hypermodeinc/modus/sdk/go/pkg/models/openai"
)

type ChatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type ResponseWithLogs struct {
	Response    string   `json:"response"`
	Role        string   `json:"role"`
	Logs        []string `json:"logs"`
	ChatHistory string   `json:"chat_history"`
}

func llmWithTools(
	model *openai.ChatModel,
	tools []openai.Tool,
	systemPrompt string,
	question string,
	chatHistory string,
	responseFormat openai.ResponseFormat,
	toolCallback func(toolCall openai.ToolCall) (*string, error),
	limit int,
	historyLength int,
) ResponseWithLogs {
	var chatMessages []openai.RequestMessage
	if chatHistory != "" {
		var err error
		chatMessages, err = openai.ParseMessages([]byte(chatHistory))
		if err != nil {
			fmt.Println("Error parsing chat history:", err)
			chatMessages = []openai.RequestMessage{}
		}
	} else {
		chatMessages = []openai.RequestMessage{}
	}

	logs := []string{}
	var finalResponse *openai.CompletionMessage
	loops := 0

	for loops < limit {
		message, _ := getLLMResponse(model, tools, systemPrompt, question, chatMessages, responseFormat)
		if loops == 0 {
			chatMessages = append(chatMessages, openai.NewUserMessage(question))
		}
		if len(message.ToolCalls) > 0 {
			chatMessages = append(chatMessages, message.ToAssistantMessage())
			for _, toolCall := range message.ToolCalls {
				logs = append(logs, fmt.Sprintf("Calling function: %s with %s", toolCall.Function.Name, toolCall.Function.Arguments))
				content, err := toolCallback(toolCall)
				if err != nil {
					fmt.Printf("Error executing tool: %+v\n", err)
					msg := fmt.Sprintf("Error executing tool: %+v , explain to the user.", err)
					chatMessages = append(chatMessages, openai.NewToolMessage(&msg, toolCall.Id))
					break
				} else {
					fmt.Println("Tool response:", content)
					chatMessages = append(chatMessages, openai.NewToolMessage(content, toolCall.Id))
				}
			}
		} else {
			chatMessages = append(chatMessages, message.ToAssistantMessage())
			finalResponse = &message
			break
		}
		loops++
	}
	// limit the chat history
	if len(chatMessages) > historyLength {
		chatMessages = chatMessages[len(chatMessages)-historyLength:]
	}

	chatHistoryJSON, _ := json.Marshal(chatMessages)

	return ResponseWithLogs{
		Response: func() string {
			if finalResponse != nil {
				return finalResponse.Content
			} else {
				return "no response"
			}
		}(),
		Role:        "assistant",
		ChatHistory: string(chatHistoryJSON),
		Logs:        logs,
	}
}

func getLLMResponse(
	model *openai.ChatModel,
	tools []openai.Tool,
	systemPrompt string,
	question string,
	chatMessages []openai.RequestMessage,
	responseFormat openai.ResponseFormat,
) (openai.CompletionMessage, error) {

	input, _ := model.CreateInput(
		openai.NewSystemMessage(systemPrompt),
	)

	input.Messages = append(input.Messages, chatMessages...)
	input.Messages = append(input.Messages, openai.NewUserMessage(question))

	input.ResponseFormat = responseFormat

	if len(tools) > 0 {
		input.Tools = tools
		input.ToolChoice = openai.ToolChoiceAuto
	}
	output, err := model.Invoke(input)
	if err != nil {
		return openai.CompletionMessage{}, err
	}

	return output.Choices[0].Message, nil
}

func main() {
	fmt.Println("Go implementation using Modus SDK")
}
